Add `--no-pager` flag

cristiand391 opened the pull request:
This PR adds a new (hidden) global flag `--no-pager` to allow disabling the pager in all sub-commands.

closes #3617

cristiand391 committed:
Add `--no-pager` flag

mislav reviewed (changes requested):
Thanks for starting on this!

I would much more prefer to see it as a global flag. You can add one to PersistentFlags on the root command (`NewCmdRoot()`) and add a PersistentPreRun hook that reads the value of the flag and calls `f.IOStreams.SetPager()` accordingly.

mgaffney commented:
+1 for making it a global flag. It would also be great if it supported the short option of `-P` like [`git`](https://git-scm.com/docs/git#Documentation/git.txt--P) does.

cristiand391 commented:
> I would much more prefer to see it as a global flag. You can add one to PersistentFlags on the root command (NewCmdRoot()) and add a PersistentPreRun hook that reads the value of the flag and calls f.IOStreams.SetPager() accordingly.

Totally missed that, thanks!

Adding it to `NewCmdRoot` will make `gh` show the flag in commands that doesn't even use the pager, what do you think about moving this functionality to its own file (like [repo_override](https://github.com/cli/cli/blob/trunk/pkg/cmdutil/repo_override.go)) and then enable it per subcommand? 

> It would also be great if it supported the short option of -P like git does.

@mgaffney I avoided adding a shorthand since IMO it's not something users will commonly use (I guess most will just alias it) and also to leave `-P` short flag free to use in some future command. What do you think @mislav ?

mislav commented:
> Adding it to `NewCmdRoot` will make `gh` show the flag in commands that doesn't even use the pager, what do you think about moving this functionality to its own file (like [repo_override](https://github.com/cli/cli/blob/trunk/pkg/cmdutil/repo_override.go)) and then enable it per subcommand?

That sounds great to me! Good point about `--no-pager` being confusing to see in help of commands that don't make use of a pager at all. 👍 

An alternative could be to still add `--no-pager` as a global flag on root, but mark it as hidden. That way it doesn't show up in help text anywhere, but we could still separately point out that it exists. You are free to choose an approach that feels right to you. Note that we are much more keen on accepting PRs with small diffs than those with large ones. 👌 

> and also to leave `-P` short flag free to use in some future command.

I agree that "global" flags should not have shorthands so that the shorthand can be used for something else in a command. Right now we don't use `-P` for anything, but we might want to in the future.

mislav commented:
@cristiand391 Can you confirm that the added PersistentPreRun block doesn't conflict with this one added for the purpose of `-R/--repo` flag (or GH_REPO environment variable) override? https://github.com/cli/cli/blob/dee89a1b6a7a346eb5109f5c6b798ecfa7702f25/pkg/cmdutil/repo_override.go#L13-L13

cristiand391 commented:
@mislav I noticed that exactly after I force-pushed (sorry), that's why I marked this PR as draft 👍🏾.
Will update later today.

cristiand391 committed:
Make sure rootCmd's PersistentPreRun hook is executed

Some sub-commands already set a PersistentPreRun hook that overrides
the one defined in rootCmd.

The solution is just to call rootCmd's PersistentPreRun hook inside
of children's PersistentPreRun hook.

cristiand391 commented on pkg/cmdutil/repo_override.go:
```diff
@@ -11,6 +11,11 @@ func EnableRepoOverride(cmd *cobra.Command, f *Factory) {
+		// Execute rootCmd's PersistentPreRun hook
+		if rootPPreRun := cmd.Root().PersistentPreRun; rootPPreRun != nil {
+			rootPPreRun(cmd, args)
+		}
```
This is needed in order to execute rootCmd's PersistentPreRun hook on all sub-commands that use `EnableRepoOverride` to set their own hook, see https://github.com/spf13/cobra/issues/1198.

cristiand391 commented:
I've manually tested the new flag and it works on both sub-commands that set their own PersistentPreRun hook and those who don't (like `api`, `gist`, etc).

Should I add a test for this flag?

demianstack6 reviewed (commented):
Z z

cristiand391 commented:
Closing in favor of #3789.
